README
Dist::Zilla::Plugin::MetaConfig::Deep

This module serves as an experimental space for features I think the
core MetaConfig *should* provide, but in a state of uncertainty about
how they should be implemented.

The objective is to extract more metadata about plugins without plugins
having to percolate hand-written adjustments system-wide to get a useful
interface.

  Composition Data

This exposes data about the roles and parent classes, and their
respective versions in play on a given plugin, to give greater depth for
problem diagnosis.

  {
    "class" : "Dist::Zilla::Plugin::Author::KENTNL::CONTRIBUTING",
    "config" : {...},
    "name" : "@Author::KENTNL/Author::KENTNL::CONTRIBUTING",
    "version" : "0.001005",
    "x_composes" : {
      "Dist::Zilla::Plugin::GenerateFile::FromShareDir" : "0.009",
      "Dist::Zilla::Role::AfterBuild"                   : "5.041",
      "Dist::Zilla::Role::AfterRelease"                 : "5.041",
      "Dist::Zilla::Role::ConfigDumper"                 : "5.041",
      "Dist::Zilla::Role::FileGatherer"                 : "5.041",
      "Dist::Zilla::Role::FileInjector"                 : "5.041",
      "Dist::Zilla::Role::FileMunger"                   : "5.041",
      "Dist::Zilla::Role::Plugin"                       : "5.041",
      "Dist::Zilla::Role::RepoFileInjector"             : "0.005",
      "Dist::Zilla::Role::TextTemplate"                 : "5.041",
      "Moose::Object"                                   : "2.1604",
      "MooseX::SlurpyConstructor::Role::Object"         : "1.2"
    }
  }

@ETHER has already made excellent inroads into making this sort of
metadata exposed via exporting "version" in all "metaconfig" plugin's
she has access to, and this is an attempt at providing the same level of
insight without requiring so much explicit buy-in from plugin authors.

This also has the neat side effect of showing what phases a plug-in is
subscribed to.

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for Dist-Zilla-Plugin-MetaConfig-Deep

0.001002 2017-03-07T09:32:59Z bf77133

0.001001 2017-03-07T09:32:10Z 2bef685
 - Bugfix: Avoid test failures with -Ddefault_inc_excludes_dot

 [Dependencies::Stats]
 - Dependencies changed since 0.001000, see misc/*.deps* for details
 - develop: +2 ↑1 -1 (suggests: ↑2)
 - test: ↓1

0.001000 2015-11-15T03:56:08Z d6682ce
 - First version.


t/lib/T/Test.pm
use 5.006;    # utf8
use strict;
use warnings;

package T::Test;

# ABSTRACT: An abstract node representing a comparison result

# AUTHORITY

our %HAS = (
  subject       => undef,
  rule          => undef,
  result        => undef,
  result_reason => undef,
  children      => sub { [] },
);

our $METACLASS = T::Test::Meta->new( \%HAS );

sub new           { $METACLASS->CREATE(@_) }
sub subject       { $METACLASS->ATTR_GET( $_[0], 'subject' ) }
sub rule          { $METACLASS->ATTR_GET( $_[0], 'rule' ) }
sub result        { $METACLASS->ATTR_GET( $_[0], 'result' ) }
sub result_reason { $METACLASS->ATTR_GET( $_[0], 'result_reason' ) }
sub children      { @{ $METACLASS->ATTR_GET( $_[0], 'children' ) } }

sub pass {
  $METACLASS->ATTR_SET( $_[0], 'result',        1 );
  $METACLASS->ATTR_SET( $_[0], 'result_reason', $_[1] );
}

sub fail {
  $METACLASS->ATTR_SET( $_[0], 'result',        0 );
  $METACLASS->ATTR_SET( $_[0], 'result_reason', $_[1] );
}
sub ok {
  $METACLASS->ATTR_SET( $_[0], 'result',        $_[1] );
  $METACLASS->ATTR_SET( $_[0], 'result_reason', $_[2] );
}

sub new_child {
  my $child = $METACLASS->CREATE(@_);
  $METACLASS->ATTR_PUSH( $_[0], 'children', $child );
  return $child;
}

{

  package T::Test::Meta;

  use Scalar::Util qw( blessed );

  sub new { bless PARAM_HASH( $_[1] ), CLASS( $_[0] ) }

  sub CREATE {
    return bless PARAM_HASH( $_[2] ), CLASS( $_[1] );
  }

  sub PARAM_HASH {
    { %{ $_[0] || {} } }
  }
  sub CLASS { blessed $_[0] || $_[0] }
  sub INSTANCE { blessed $_[0] ? $_[0] : die "not an instance" }

  sub ATTR_SET {
    my ( $meta, $instance, $attrname, $value ) = @_;
    $instance->{$attrname} = $value;
    return $instance;
  }

  sub ATTR_GET {
    my ( $meta, $instance, $attrname ) = @_;
    if ( $meta->{$attrname} and not exists $instance->{$attrname} ) {
      $instance->{$attrname} = $meta->{$attrname}->($instance);
    }
    return $instance->{$attrname};
  }

  sub ATTR_MUTATE {
    my ( $meta, $instance, $attrname, $maybe_value ) = @_;
    if ($maybe_value) {
      $meta->ATTR_SET( $instance, $attrname, $maybe_value );
    }
    return $meta->ATTR_GET( $invocant, $attribute );
  }

  sub ATTR_PUSH {
    my ( $meta, $invocant, $attrname, $value ) = @_;
    push @{ $instance->{$attrname} }, $value;
    return $instance;
  }
}
1;



t/lib/T/Chain.pm
use 5.006;    # our
use strict;
use warnings;

package T::Chain;

# ABSTRACT: A prototype for a Test::Chain

# AUTHORITY

sub metachain {
  require T::Chain::Meta;
  T::Chain::Meta::metachain( $_[0] );
}

sub new     { $_[0]->metachain->create_instance(@_) }
sub matches { $_[0]->metachain->instance_matches(@_) }

1;



t/lib/T/HashRule.pm
use strict;
use warnings;

package T::HashRule;

use parent 'T::Chain';

use Exporter 5.57 qw(import);

our @EXPORT_OK = ('is_hash');

sub is_hash {
  if ( scalar @_ == 2 ) {
    my ( $label, $code ) = @_;
    my $rule = __PACKAGE__->new( { label => $label } );
    $rule->_is_hash;
    $code->($rule);
    return $rule;
  }
  if ( scalar @_ == 1 ) {
    my $rule = __PACKAGE__->new( { label => $_[0] } );
    $rule->_is_hash;
    return $rule;
  }
}

__PACKAGE__->metachain->add_rule(
  '_is_hash' => sub {
    return sub {
      my ( $context, $got ) = @_;
      return $context->ok( 1, 'is of reftype hash' ) if 'HASH' eq ref $got;
      return $context->ok( 0, 'is not of reftype hash' );
    };
  }
);
__PACKAGE__->metachain->add_rule(
  'has_key' => sub {
    my ( $key, $subrule ) = @_;
    if ( not $subrule ) {
      return sub {
        my ( $context, $got ) = @_;
        if ( not exists $got->{$key} ) {
          return $context->ok( 0, 'hash does not contain ' . $key );
        }
        $context->ok( 1, 'hash contains key ' . $key );
      };
    }
    if ( not $subrule->can('matches') ) {
      die "Inappropriate subrule";
    }
    return sub {
      my ( $context, $got ) = @_;
      if ( not exists $got->{$key} ) {
        return $context->ok( 0, 'hash does not contain ' . $key );
      }
      $context->ok( 1, 'hash contains key ' . $key );
      return $subrule->matches( $got->{$key}, $context );
    };
  }
);
__PACKAGE__->metachain->add_rule(
  'keys_are' => sub {
    my ($key_array) = @_;
    my $wanted_keys = {};
    @{$wanted_keys}{@_} = (1) x scalar @_;
    return sub {
      my ( $context, $got ) = @_;
      my $available = {};
      @{$available}{ keys %{$got} } = (1) x keys %{$got};
      $context->group(
        'keys_are' => sub {
          my $ret = $context->group(
            'required_keys' => sub {
              for my $key ( sort keys %{$wanted_keys} ) {
                if ( exists $available->{$key} ) {
                  delete $available->{$key};
                  $context->ok( 1, "Has required key $key" );
                  next;
                }
                return $context->ok( 0, "Lacks required key $key" );
              }
            }
          );
          return $ret if not $ret;
          if ( keys %{$available} ) {
            return $context->ok( 1, "No excess keys" );
          }
          $context->ok( 0, ( scalar keys %{$available} ) . " excess keys" );
          return 0;
        }
      );
    };
  }
);
1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Experimental enhancements to MetaConfig",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Dist-Zilla-Plugin-MetaConfig-Deep",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Dist::Zilla::Plugin::MetaConfig::Deep",
  "PREREQ_PM" => {
    "Dist::Zilla::Role::MetaProvider" => 0,
    "Moose" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.001002",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Dist::Zilla::Role::MetaProvider" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Moose" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
